### PR TITLE
Update user

### PR DESCRIPTION
**Pull request by sleepyzzzzzz**
## Changes made in this PR
add UpdateUser implementation + specs in domain and grpc layer

## Corresponding noted Ruby branch
https://github.com/ParamountDataManagement/paramount-api/tree/yiz/users-entity

## Implementation Checklist
Please check following check list to ensure the tasks are considered and completed in this branch.

#### API Design
- [x] Defined gRPC API in proto file
- [x] regenerated go code using `make proto_generate` command

#### API implementation
- [x] Implemented API in corresponding gRPC server
- [x] Added tests in gRPC layer

#### Domain Service and DB query Implementation
- [ ] Added new query in dbstore layer
- [ ] Added tests in db store layer
                                                                                       
- [x] Implemented functionality in Domain Service Layer
- [x] Added test in Domain Service Layer

- [ ] Created a new branch in paramount-api, with comments there indicating the functionality in Ruby that's re-implemented in Golang

#### Postgres Layer - identify and potentially add db migration to create index for fast querying.
- [ ] Verified and potentially created index for the newly added query.

### CODE REVIEW

File: apis/grpcservice/grpcusers/usersGRPCServer_test.go [hunk 1]
```diff
@@ -650,53 +651,91 @@ var _ = Describe("UsersGRPCServer", func() {
...
+					IsActive:  true,
+					FirstName: "Lu",
+					LastName:  "Han",
+					Email:     testUserEmail + testHelpIndex,
+					Password:  password,
+					AccountID: testAccount.ID,
+				}
+				resultUser, err = usersStore.CreateUser(context.Background(), testUserInfo)
+				Expect(err).To(BeNil())
+				_, err = db.Exec(`insert into user_access_tokens(user_id, token, expires_at, user_agent, user_agent_encrypted)
```
Comment (niki-joanne) [inline apis/grpcservice/grpcusers/usersGRPCServer_test.go:687]: Can we add a comment specifying what this is meant to simulate?

File: apis/grpcservice/grpcusers/usersGRPCServer_test.go [hunk 2]
```diff
@@ -650,53 +651,91 @@ var _ = Describe("UsersGRPCServer", func() {
...
+							IsActive:        resultUser.IsActive,
+							FirstName:       resultUser.FirstName,
+							LastName:        resultUser.LastName,
+							Email:           resultUser.Email,
+							Password:        resultUser.Password,
+							IsAdministrator: resultUser.IsAdministrator,
+							AccountId:       resultUser.AccountID,
+							IsPdmUser:       true,
+							ShouldSendEmail: resultUser.ShouldSendEmail,
+						},
```
Comment (niki-joanne) [inline apis/grpcservice/grpcusers/usersGRPCServer_test.go:712]: Can we just use the `UserToProto()` function here, with `resultUser` as the argument?

File: apis/grpcservice/grpcusers/usersGRPCServer_test.go (hunk 1 above)
Comment (sleepyzzzzzz) [reply apis/grpcservice/grpcusers/usersGRPCServer_test.go:689]: added

File: internal/domain/users/domainService.go [hunk 3]
```diff
@@ -330,6 +330,37 @@ func (s *DomainService) confirm(
...
 
+// UpdateUser updates the record in users table and return the updated user info
+func (s *DomainService) UpdateUser(
+	ctx context.Context,
+	updateInfoFields []string,
+	user UserInfo,
+	userToken string,
+	userAgent string,
+) (*UserInfo, error) {
+	// deactivate all user access tokens by user id
```
Comment (lhan-the-real) [inline internal/domain/users/domainService.go:341]: In Ruby, it first checks if the updateUserInfo's email address is not empty, ensuring the email address doesn't belong to any existing user. Can we add this logic?

File: apis/grpcservice/grpcusers/usersGRPCServer.go [hunk 4]
```diff
@@ -553,6 +553,47 @@ func (s *usersGRPCServer) GetUser(
 	return &usersv1.GetUserResponse{User: grpcutils.UserToProto(user)}, nil
 }
 
+// UpdateUser updates an user with given fields and update data
+func (s *usersGRPCServer) UpdateUser(
+	ctx context.Context,
+	request *usersv1.UpdateUserRequest,
+) (*usersv1.UpdateUserResponse, error) {
+	if request.GetUserToken() == "" {
```
Comment (lhan-the-real) [inline apis/grpcservice/grpcusers/usersGRPCServer.go:561]: Let's add a token validation in gRPC Server layer

File: internal/domain/users/usersStore.go [hunk 5]
```diff
@@ -37,6 +37,7 @@ type UsersStore interface {
 	GetUserValidToken(ctx context.Context, userID int32, userAgent string) (string, bool, error)
 	ValidateToken(ctx context.Context, userToken string, userAgent string) (int32, error)
 	InvalidateToken(ctx context.Context, accessToken string, userAgent string) error
+	InvalidateTokenByID(ctx context.Context, id int32) error
```
[the remark this answers is not in the document]
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore.go:40]: Yes. idx created
1. user_id
2. token + user_agent_encrypted

File: internal/domain/users/usersStore.go [hunk 6]
```diff
@@ -235,7 +230,15 @@ func (u *usersStore) UpdateUser(
 		updateInfoFields,
 		editableUserInfoFields,
 	); err != nil {
-		return UserInfo{}, pdmerr.NewParent(err, " Update User, invalid input")
+		return UserInfo{}, pdmerr.NewParent(err, "Update User, invalid input")
```
[the remark this answers is not in the document]
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore.go:233]: fix

File: internal/domain/users/usersStore.go [hunk 7]
```diff
@@ -46,19 +47,13 @@ type UsersStore interface {
...
-	"password":               true,
-	"confirmation_token":     true,
-	"confirmed_at":           true,
-	"verification_token":     true,
-	"verified_at":            true,
-	"reset_token":            true,
-	"reset_token_expired_at": true,
-	"is_administrator":       true,
-	"account_id":             true,
+	"is_active":         true,
```
[the remark this answers is not in the document]
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore.go:50]: looking through the ruby code now
didn't find there is intercom_user has email updates, could you give me a file that you saw there is an update on eamil?

File: internal/domain/users/usersStore_test.go [hunk 8]
```diff
@@ -679,6 +687,26 @@ var _ = Describe("UsersStore", func() {
...
+
+			DescribeTable("InvalidateTokenByID", func(
+				inputUserID int32,
+				expectedHasError bool,
+			) {
+				err := usersStore.InvalidateTokenByID(context.Background(), inputUserID)
+				if expectedHasError {
+					Expect(err).To(HaveOccurred())
+				} else {
+					Expect(err).To(BeNil())
```
[the remark this answers is not in the document]
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore_test.go:699]: added

File: internal/domain/users/usersStore_test.go [hunk 9]
```diff
@@ -679,6 +687,26 @@ var _ = Describe("UsersStore", func() {
...
+				}
+			},
+				Entry("should return error with invalid user id", int32(-1), true),
```
[the remark this answers is not in the document]
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore_test.go:702]: added

File: internal/domain/users/domainService.go (hunk 3 above)
Comment (sleepyzzzzzz) [reply internal/domain/users/domainService.go:341]: added

File: internal/domain/users/usersStore_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore_test.go:701]: But currently there doesn't exist a function to check if the timestamp changes

File: internal/domain/users/usersStore.go [hunk 10]
```diff
@@ -476,6 +483,33 @@ func (u *usersStore) InvalidateToken(ctx context.Context, accessToken string, us
 	return nil
 }
 
+// InvalidateTokenByID sets the user's token to expire now by the user id
+func (u *usersStore) InvalidateTokenByID(ctx context.Context, id int32) error {
+	query := `update user_access_tokens
```
[the remark this answers is not in the document]
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore.go]: fix

File: internal/domain/users/usersStore_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore_test.go:698]: added

File: apis/grpcservice/grpcusers/usersGRPCServer.go (hunk 4 above)
Comment (sleepyzzzzzz) [reply apis/grpcservice/grpcusers/usersGRPCServer.go:561]: added

File: internal/domain/users/usersStore.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore.go:50]: add todo

[the remark this answers is not in the document]
Comment (lhan-the-real) [reply internal/domain/users/usersStore.go:50]: Let's create a separate clickup task  to track that effort.

File: internal/domain/users/usersStore_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (lhan-the-real) [reply internal/domain/users/usersStore_test.go:701]: we can use GetValidToken() to expect returning none valid token

File: internal/domain/users/usersStore.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore.go:50]: added

File: internal/domain/users/usersStore_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore_test.go:701]: added

File: internal/domain/users/usersStore_test.go [hunk 11]
```diff
@@ -679,6 +695,73 @@ var _ = Describe("UsersStore", func() {
...
+
+					_, found, err = usersStore.GetUserValidToken(
+						context.Background(),
+						expectedUserInfo.ID,
+						"new user agent",
+					)
+					Expect(found).To(BeFalse())
+					Expect(err).To(BeNil())
+				})
+			})
```
Comment (lhan-the-real) [inline internal/domain/users/usersStore_test.go:764]: Does this testing multiple valid token for the same user id are invalidated? Seems there's only 1 valid token created in the test function.

File: internal/domain/users/usersStore.go [hunk 12]
```diff
@@ -448,17 +459,16 @@ func (u *usersStore) ValidateToken(ctx context.Context, userToken string, userAg
 	return userID, nil
 }
 
-// InvalidateToken sets the user's token to expire now
+// InvalidateToken sets the user's token to expire now by token and agent
```
Comment (lhan-the-real) [inline internal/domain/users/usersStore.go:462]: Let's update the comment to `the token to be deleted`

File: internal/domain/users/usersStore.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (lhan-the-real) [reply internal/domain/users/usersStore.go:50]: Could you add a link to the Clickup Task here? So I can follow up on that.

File: apis/grpcservice/grpcusers/usersGRPCServer_test.go [hunk 13]
```diff
@@ -636,6 +648,140 @@ var _ = Describe("UsersGRPCServer", func() {
...
+							UserToken:  "test access token",
+							UserAgent:  testUserAgent,
+							User:       user,
+							UpdateMask: updateInfoFields,
+						},
+					)
+					Expect(err).To(HaveOccurred())
+				})
+
+				It("return updated user", func() {
```
Comment (lhan-the-real) [inline apis/grpcservice/grpcusers/usersGRPCServer_test.go:765]: Do we have test against fields that cannot be updated?

File: internal/domain/users/usersStore.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore.go:50]: https://app.clickup.com/t/2ccufnd

File: internal/domain/users/usersStore_test.go (hunk 11 above)
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore_test.go:764]: Yes. In BeforeEach, there is already a access token with testAgent inserted to the db

File: internal/domain/users/usersStore.go (hunk 12 above)
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore.go:462]: fix

File: apis/grpcservice/grpcusers/usersGRPCServer_test.go (hunk 13 above)
Comment (sleepyzzzzzz) [reply apis/grpcservice/grpcusers/usersGRPCServer_test.go:765]: yes, starting from line 744

File: internal/domain/users/usersStore_test.go (hunk 11 above)
Comment (lhan-the-real) [reply internal/domain/users/usersStore_test.go:764]: I see. Thanks. I was searching it, but didn't find it. Thanks for pointing out.

**Review by lhan-the-real — APPROVED** [review]

**Pull request merged**